### PR TITLE
Fix and speed-up AudioDecoder seeking logic

### LINKED ISSUE #1448: `AudioDecoder.get_samples_played_in_range` scales linearly with `start_seconds`
### 🐛 Describe the bug

Related to https://github.com/meta-pytorch/torchcodec/issues/1429, `AudioDecoder.get_samples_played_in_range` appears to decode the file from the beginning up to `start_seconds`, rather than seeking before decoding. As a result, decoding a short window near the end of a long file is roughly as expensive as a full decode. For comparison, `soundfile.SoundFile.seek(...) + read(...)` stays effectively constant-time regardless of `start_seconds`.

I tested this with wav, flac, ogg and mp3 files, and the behavior is the same (apart from soundfile not being constant anymore for mp3).

Short script in the details block for replication:
<details>

```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "numpy>=2.4.6",
#     "soundfile>=0.13.1",
#     "torch>=2.12.0",
#     "torchaudio>=2.11.0",
#     "torchcodec>=0.13.0",
# ]
# ///

import argparse
import time
from pathlib import Path
from statistics import mean, stdev

import soundfile as sf
from torchcodec.decoders import AudioDecoder


def bench(label: str, fn, n: int = 5) -> None:
    times = []
    for _ in range(n):
        t0 = time.perf_counter()
        fn()
        times.append(time.perf_counter() - t0)
    avg = mean(times)
    sd = stdev(times) if len(times) > 1 else 0.0
    print(f"  {label:<55s} {avg * 1000:8.2f} ms  ± {sd * 1000:.2f} ms")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("path", type=Path, help="Path to the audio file to benchmark")
    parser.add_argument(
        "-w",
        "--window",
        type=float,
        default=1.0,
        help="Window size in seconds for the partial-decode benchmark (default: 1.0)",
    )
    args = parser.parse_args()
    file_path: Path = args.path
    window: float = args.window

    with sf.SoundFile(str(file_path)) as f:
        sr = f.samplerate
        n_frames = f.frames
        duration = n_frames / sr
    print(f"File: {file_path.name}")
    print(f"Sample rate: {sr} Hz, frames: {n_frames}, duration: {duration:.1f} s\n")

    print("== Full decode ==")
    bench("soundfile sf.read (full)", lambda: sf.read(str(file_path)))
    bench(
        "torchcodec AudioDecoder.get_all_samples (full)",
        lambda: AudioDecoder(str(file_path)).get_all_samples(),
    )

    print(f"\n== Decode a {window:g}-second window at various start offsets ==")
    for start in [0.0, duration * 0.25, duration * 0.5, duration * 0.75, max(0.0, duration - window - 1.0)]:
        stop = min(start + window, duration)
        print(f"\n-- start = {start:7.2f} s  (stop = {stop:7.2f} s) --")

        def sf_seek_read(start=start, stop=stop):
            with sf.SoundFile(str(file_path)) as f:
                f.seek(int(start * sr))
                return f.read(int((stop - start) * sr))

        def tc_range(start=start, stop=stop):
            dec = AudioDecoder(str(file_path))
            return dec.get_samples_played_in_range(start_seconds=start, stop_seconds=stop)

        bench(f"soundfile seek + read {window:g} s window", sf_seek_read)
        bench(f"torchcodec get_samples_played_in_range {window:g} s window", tc_range)


if __name__ == "__main__":
    main()
```
</details>

Output of this script when run against a long file from VoxPopuli:
```txt
❯ uv run issue.py 20090112-0900-PLENARY-10_fr.ogg -w 10
File: 20090112-0900-PLENARY-10_fr.ogg
Sample rate: 16000 Hz, frames: 44037781, duration: 2752.4 s

== Full decode ==
  soundfile sf.read (full)                                  580.63 ms  ± 17.76 ms
  torchcodec AudioDecoder.get_all_samples (full)            434.91 ms  ± 12.32 ms

== Decode a 10-second window at various start offsets ==

-- start =    0.00 s  (stop =   10.00 s) --
  soundfile seek + read 10 s window                           2.32 ms  ± 0.16 ms
  torchcodec get_samples_played_in_range 10 s window          1.85 ms  ± 0.15 ms

-- start =  688.09 s  (stop =  698.09 s) --
  soundfile seek + read 10 s window                           2.33 ms  ± 0.09 ms
  torchcodec get_samples_played_in_range 10 s window         66.03 ms  ± 0.57 ms

-- start = 1376.18 s  (stop = 1386.18 s) --
  soundfile seek + read 10 s window                           2.30 ms  ± 0.11 ms
  torchcodec get_samples_played_in_range 10 s window        131.11 ms  ± 1.52 ms

-- start = 2064.27 s  (stop = 2074.27 s) --
  soundfile seek + read 10 s window                           2.46 ms  ± 0.18 ms
  torchcodec get_samples_played_in_range 10 s window        196.03 ms  ± 2.29 ms

-- start = 2741.36 s  (stop = 2751.36 s) --
  soundfile seek + read 10 s window                           2.33 ms  ± 0.11 ms
  torchcodec get_samples_played_in_range 10 s window        270.86 ms  ± 5.58 ms
```

### Versions

Collecting environment information...
PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.7.4 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: version 4.3.2
Libc version: N/A

Python version: 3.12.13 (main, Apr 14 2026, 14:33:40) [Clang 22.1.3 ] (64-bit runtime)
Python platform: macOS-15.7.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.13.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by NicolasHug**
Our seeking logic has been to always seek back to the beginning of the file, to avoid frame dependencies issues: on lossy formats, frame i - 1 usually needs to be decoded for frame i to be properly decoded. Upon seek, we flush (as we should!), and we start by decoding the target frame i without having decoded frame i-1, leading to incorrect output.

The workaround we had for that was to always seek back at the beginning of the file. Which is obviously suboptimal in many scenarios.

This PR avoids this costly workaround by always seeking to frame i - n (usually n = 4). I'm more confident with doing that now that I could verify that other libraries do the same trick. See comments in the code for more details.

Closes https://github.com/meta-pytorch/torchcodec/issues/1448

See also https://github.com/meta-pytorch/torchcodec/issues/1429#issuecomment-4539291978 and https://github.com/meta-pytorch/torchcodec/pull/507#issuecomment-3504708283 for related discussions.

#### Benchmarks

Decoding the last 10s of a 10m mp3 file:

```
# On main:
med = 228.57ms +- 1.77
med = 226.73ms +- 24.94

# on this PR:
med = 4.49ms +- 0.15
med = 4.61ms +- 0.18
```

```py
import torch
from time import perf_counter_ns
from torchcodec.decoders import AudioDecoder


def bench(f, *args, num_exp=100, warmup=0, **kwargs):
    for _ in range(warmup):
        f(*args, **kwargs)
    times = []
    for _ in range(num_exp):
        start = perf_counter_ns()
        f(*args, **kwargs)
        end = perf_counter_ns()
        times.append(end - start)
    return torch.tensor(times).float()


def report_stats(times, unit="ms"):
    mul = {
        "ns": 1,
        "µs": 1e-3,
        "ms": 1e-6,
        "s": 1e-9,
    }[unit]
    times = times * mul
    std = times.std().item()
    med = times.median().item()
    print(f"{med = :.2f}{unit} +- {std:.2f}")
    return med


decoder = AudioDecoder("bench_10min.mp3")
times = bench(decoder.get_samples_played_in_range, 590, 600, num_exp=100, warmup=5)
report_stats(times)
```

```
ffmpeg -f lavfi -i "sine=frequency=440:duration=600" -codec:a libmp3lame -b:a 128k bench_10min.mp3 -y
```

### CODE REVIEW

**Comment by NicolasHug** [comment]
Not strictly speaking a bug fix, but I'll label it as such for the release notes.
